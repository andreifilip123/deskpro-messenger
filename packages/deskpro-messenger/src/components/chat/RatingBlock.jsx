import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { injectIntl } from 'react-intl';

import Button from '../form/Button';
import ChatPrompt from '../ui/ChatPrompt';

const transMessages = {
  questionHeader: {
    id: 'chat.rating_block.question_header',
    defaultMessage: 'Rate your conversation with {name}'
  },
  buttonHelpful: {
    id: 'chat.rating_block.buttons.helpful',
    defaultMessage: 'Helpful'
  },
  buttonUnhelpful: {
    id: 'chat.rating_block.buttons.unhelpful',
    defaultMessage: 'Unhelpful'
  },
  thankYouHeader: {
    id: 'chat.rating_block.thank_you_header',
    defaultMessage: 'Thank you for your feedback'
  }
};

class RatingBlock extends PureComponent {
  static propTypes = {
    onSend: PropTypes.func.isRequired,
    message: PropTypes.shape({
      type: PropTypes.string.isRequired,
      origin: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired
  };

  handleClick = (e) => {
    this.props.onSend({
      ...this.props.message,
      origin: 'user',
      rate: e.target.name === 'helpful'
    });
  };

  render() {
    const { rate } = this.props.message;

    return typeof rate === 'undefined'
      ? this.renderButtons()
      : this.renderSmile();
  }

  renderButtons() {
    const { intl, message } = this.props;
    return (
      <ChatPrompt
        header={intl.formatMessage(transMessages.questionHeader, message)}
      >
        <div className="dpmsg-PromptContentEvaluation">
          <Button
            width="limited"
            color="success"
            name="helpful"
            onClick={this.handleClick}
          >
            {intl.formatMessage(transMessages.buttonHelpful)}
          </Button>
          <Button
            width="limited"
            color="danger"
            name="unhelpful"
            onClick={this.handleClick}
          >
            {intl.formatMessage(transMessages.buttonUnhelpful)}
          </Button>
        </div>
      </ChatPrompt>
    );
  }

  renderSmile() {
    const { rate } = this.props.message;
    return (
      <ChatPrompt
        header={this.props.intl.formatMessage(transMessages.thankYouHeader)}
      >
        <div className="dpmsg-PromptContentEvaluation">
          <i
            className={classNames(
              {
                'dpmsg-IconSmile': rate,
                'dpmsg-Icon': !!rate
              },
              'is-blue'
            )}
          />
        </div>
      </ChatPrompt>
    );
  }
}

export default injectIntl(RatingBlock);
