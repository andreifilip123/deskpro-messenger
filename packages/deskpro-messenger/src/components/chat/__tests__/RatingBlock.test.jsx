import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from 'react-testing-library';

import { RatingBlock } from '../RatingBlock';

describe('<RatingBlock />', () => {
  const intl = {
    formatMessage: jest.fn(({ id }) => id)
  };

  it('should render correctly', () => {
    const onSend = jest.fn();
    const tree = renderer
      .create(<RatingBlock intl={intl} onSend={onSend} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should send message and update state on rating selection', () => {
    const onSend = jest.fn();
    const { getByText } = render(<RatingBlock intl={intl} onSend={onSend} />);
    expect(getByText('chat.rating_block.question_header')).toBeInTheDocument();
    fireEvent.click(getByText('chat.rating_block.buttons.helpful'));
    expect(onSend).toBeCalledWith({
      type: 'chat.rating',
      origin: 'user',
      rate: true
    });
    expect(getByText('chat.rating_block.thank_you_header')).toBeInTheDocument();
  });
});
