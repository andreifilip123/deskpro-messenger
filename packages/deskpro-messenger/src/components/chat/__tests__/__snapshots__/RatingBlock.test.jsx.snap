// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingBlock /> should render correctly 1`] = `
<div
  className="dpmsg-MessagePrompt"
>
  <div
    className="dpmsg-PromptHeader"
  >
    <span
      className="dpmsg-PromptHeaderText"
    >
      chat.rating_block.question_header
    </span>
  </div>
  <div
    className="dpmsg-PromptContentEvaluation"
  >
    <button
      className="dpmsg-Button Button-LimitedWidth Button--success"
      name="helpful"
      onClick={[Function]}
    >
      chat.rating_block.buttons.helpful
    </button>
    <button
      className="dpmsg-Button Button-LimitedWidth Button--danger"
      name="unhelpful"
      onClick={[Function]}
    >
      chat.rating_block.buttons.unhelpful
    </button>
  </div>
</div>
`;
